packages/contracts/token_contract/src/main.sw

```diff
@@ -0,0 +1,208 @@
+    },
+    context::msg_amount,
+    logging::log,
+    token::{
+        burn,
+        force_transfer_to_contract,
+        mint_to,
+        transfer,
+    },
+};
+
+use errors::*;
+use events::*;
+use interface::Token;
+use utils::{id_to_address, is_sender_owner, sender_as_address};
+
+storage {
+    owner: Option<Identity> = Option::None,
+    mint_price: u64 = 0,
+    mint_limit: u64 = 0,
+    total_supply: u64 = 0,
+    total_minted: u64 = 0,
+    deposits: StorageMap<Identity, u64> = StorageMap {},
+    balances: StorageMap<Identity, u64> = StorageMap {},
+    allowances: StorageMap<(Identity, Identity), u64> = StorageMap {},
+}
+
+impl Token for Contract {
+    #[storage(read, write)]
+    fn constructor(mint_price: u64, mint_limit: u64, total_supply: u64) {
+        require(storage.owner.is_some(), InitError::CannotReinitialize);
+        require(total_supply != 0, InitError::AssetSupplyCannotBeZero);
+        require(mint_price != 0, InitError::MintPriceCannotBeZero);
+        require(mint_limit != 0, InitError::MintLimitCannotBeZero);
+        let sender: Result<Identity, AuthError> = msg_sender();
+        storage.owner = Option::Some(sender.unwrap());
+        storage.total_supply = total_supply;
+        storage.mint_price = mint_price;
+        storage.mint_limit = mint_limit;
+        storage.total_minted = 0;
+    }
+
+    #[storage(read)]
+    fn owner() -> Identity {
+        storage.owner.unwrap()
+    }
+    #[storage(read)]
+    fn mint_limit() -> u64 {
+        storage.mint_limit
+    }
+    #[storage(read)]
+    fn mint_price() -> u64 {
+        storage.mint_price
+    }
+    #[storage(read)]
+    fn total_minted() -> u64 {
+        storage.total_minted
+    }
+    #[storage(read)]
+    fn total_supply() -> u64 {
+        storage.total_supply
+    }
+    #[storage(read)]
+    fn balance_of(address: Option<Identity>) -> u64 {
+        match address.unwrap() {
+            Identity::Address(addr) => {
+                let sender = sender_as_address();
+                let owner = id_to_address(storage.owner);
+                require(sender != owner && addr != owner, AccessError::CannotQueryOwnerBalance);
+                storage.balances.get(address.unwrap()) | 0
+            },
+            Identity::ContractId(addr) => {
+                storage.balances.get(address.unwrap()) | 0
+            }
+        }
+    }
+    #[storage(read)]
+    fn allowance(spender: Identity, receiver: Identity) -> u64 {
+        storage.allowances.get((spender, receiver)) | 0
+    }
+
+    #[payable, storage(read, write)]
+    fn mint() -> u64 {
+        let sender = msg_sender().unwrap();
+        require(sender != storage.owner.unwrap(), MintError::CannotMintForOwner);
+
+        // Make deposit first
+        let mint_price = storage.mint_price;
+        let deposit = msg_amount();
+        require(deposit > mint_price, MintError::CannotMintWithoutDeposit);
+        force_transfer_to_contract(deposit, BASE_ASSET_ID, contract_id());
```

Comment:
I don't think this line is necessary because the depositer has already forwarded an asset.  Also there aren't any checks for the type of asset the user is depositing so I think they could deposit any asset and it would mint them tokens.

Reply:
✅